src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py

```diff
@@ -243,29 +253,552 @@ def apply(self, node, options=None):
+
+        :param kernel: the kernel to insert.
+        :type kernel: :py:class:`psyclone.psyir.nodes.KernelSchedule`
+        '''
+
+        if not self._kernels_file:
+            self._kernels_file = FileContainer("opencl_kernels")
+
+        # Create a copy of the kernel and remove precision symbols since they
+        # are not supported in the OpenCL backend.
+        kernel_copy = kernel.get_kernel_schedule().copy()
+        symtab = kernel_copy.symbol_table
+
+        for sym in symtab.datasymbols:
+            # Not all types have the 'precision' attribute (e.g. DeferredType)
+            if (hasattr(sym.datatype, "precision") and
+                    isinstance(sym.datatype.precision, DataSymbol)):
+                sym.datatype._precision = ScalarType.Precision.DOUBLE
+
+        if 'go_wp' in symtab:
+            del symtab._symbols['go_wp']
+
+        # Insert kernel in the OpenCL kernels file if it doesn't already exist
+        for routine in self._kernels_file.walk(Routine):
+            if routine.name == kernel.name:
+                break  # if it exist re-use existing one
+                # TODO 1572: Here we assume that in the same Invoke (scope) a
+                # kernel with the same name will be the same kernel, but than
+                # may not be true when doing multiple invokes.
+        else:
+            self._kernels_file.addchild(kernel_copy)
+
+    def _output_opencl_kernels_file(self):
+        ''' Write the current OpenCL kernels PSyIR tree into a file. '''
+
+        # TODO 1013: The code below duplicates some logic of the CodedKern
+        # rename_and_write method. Ideally this should be moved out of
+        # the AST and transformations and put into some kind of IOManager.
+
+        ocl_writer = OpenCLWriter(kernels_local_size=64)
+        new_kern_code = ocl_writer(self._kernels_file)
+
+        fdesc = None
+        name_idx = -1
+        while not fdesc:
+            name_idx += 1
+            new_name = f"opencl_kernels_{name_idx}.cl"
+
+            try:
+                # Atomically attempt to open the new kernel file (in case
+                # this is part of a parallel build)
+                fdesc = os.open(
+                    os.path.join(Config.get().kernel_output_dir, new_name),
+                    os.O_CREAT | os.O_WRONLY | os.O_EXCL)
+            except (OSError, IOError):
+                # The os.O_CREATE and os.O_EXCL flags in combination mean
+                # that open() raises an error if the file exists
+                continue
+
+        # Write the modified AST out to file
+        os.write(fdesc, new_kern_code.encode())
+        # Close the new kernel file
+        os.close(fdesc)
+
+    @staticmethod
+    def _generate_set_args_call(kernel, scope):
+        '''
+        Generate the Call statement to the set_args subroutine for the
+        provided kernel.
+
+        :param parent: parent subroutine in f2pygen AST of generated code.
```

Comment:
Please document the arguments and return value.

Reply:
Done